File: OpenBurnBarCore/Tests/OpenBurnBarCoreTests/MemorySecretPIIGateTests.swift
// SPDX-License-Identifier: AGPL-3.0-only
//
// Tests for the shared, fail-closed secret/PII gate promoted in PR-C1. Covers
// the five must-fixes: finding id+label carriage, mandatory Luhn for credit
// cards, bounded IPv4 octets, bounded fail-closed redaction, and availability.

import Foundation
import XCTest
@testable import OpenBurnBarCore
@testable import OpenBurnBarKernel

final class MemorySecretPIIGateTests: XCTestCase {
    // MARK: - Availability (must-fix #2: Bundle.module flat loader resolves)

    func testGateIsAvailableFromBundleModule() {
        // The corpus must load via Bundle.module (flat filename, no subdirectory)
        // or the filesystem fallback. A fail-closed gate would silently reject
        // everything, so this is the canary for resource wiring in Core.
        XCTAssertTrue(MemorySecretPIIGate.isAvailable, "Gate corpus failed to load — gate is fail-closed.")
        XCTAssertFalse(MemorySecretPIIGate.corpusVersion.isEmpty)
    }

    // MARK: - Positive secret detection

    func testDetectsNamedSecretsWithStableIDsAndLabels() {
        let anthropicKey = ["sk", "ant", String(repeating: "a", count: 26)].joined(separator: "-")
        let verdict = MemorySecretPIIGate.evaluate("User pasted \(anthropicKey) here.", policy: .reject)
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected reject for a secret-bearing body, got \(verdict)")
            return
        }
        XCTAssertEqual(findings.map(\.id), ["anthropic-api-key"])
        XCTAssertEqual(findings.first?.label, "Anthropic API key detected")
        XCTAssertEqual(findings.first?.kind, .secret)
    }

    func testOpenAIKeyDetectedEvenWhenEmbeddedAfterUnderscore() {
        // The chat migration contract requires `apiKey_sk-...` to still reject as
        // an OpenAI key (the corpus openai regex drops the leading \b for this).
        let body = "Embedded key apiKey_sk-abcdefghijklmnopqrstuvwxyz1234567890 must reject."
        XCTAssertEqual(MemorySecretPIIGate.findingIDs(in: body), ["openai-api-key"])
    }

    func testAnthropicKeyDoesNotAlsoTripOpenAIPattern() {
        // Negative-lookahead in the openai regex keeps a single named finding for
        // an anthropic key (chat asserts a single-element array).
        let anthropicKey = "sk-ant-1234567890abcdef1234567890"
        XCTAssertEqual(MemorySecretPIIGate.findingIDs(in: "key \(anthropicKey)"), ["anthropic-api-key"])
    }

    // MARK: - Positive PII detection

    func testDetectsEmailAndSSNAsPII() {
        let emailFindings = MemorySecretPIIGate.findings(in: "reach me at person@example.com")
        XCTAssertEqual(emailFindings.map(\.id), ["email-address"])
        XCTAssertEqual(emailFindings.first?.kind, .pii)

        let ssnFindings = MemorySecretPIIGate.findings(in: "ssn 123-45-6789")
        XCTAssertEqual(ssnFindings.map(\.id), ["us-ssn"])
        XCTAssertEqual(ssnFindings.first?.kind, .pii)
    }

    // MARK: - Negative (clean text allowed)

    func testCleanTextIsAllowed() {
        XCTAssertEqual(MemorySecretPIIGate.evaluate("I prefer dark mode and tabs over spaces."), .allow)
        XCTAssertTrue(MemorySecretPIIGate.findings(in: "Nothing sensitive here at all.").isEmpty)
    }

    func testEmptyTextIsAllowed() {
        XCTAssertEqual(MemorySecretPIIGate.evaluate(""), .allow)
    }

    // MARK: - Luhn (must-fix #3): true-positive accepts, false-positive preserved

    func testLuhnValidCreditCardIsDetected() {
        // 4111 1111 1111 1111 is the canonical Luhn-valid Visa test number.
        let findings = MemorySecretPIIGate.findings(in: "card 4111 1111 1111 1111 on file")
        XCTAssertEqual(findings.map(\.id), ["credit-card-number"])
        XCTAssertEqual(findings.first?.kind, .pii)
    }

    func testLuhnInvalidSixteenDigitRunIsNotACard() {
        // 1234 5678 9012 3456 has 16 digits but fails Luhn — must NOT be deleted as
        // a card (this is the data-integrity regression the must-fix prevents).
        let verdict = MemorySecretPIIGate.evaluate("order 1234 5678 9012 3456 shipped", policy: .reject)
        XCTAssertEqual(verdict, .allow)
    }

    func testLuhnHelperDirectly() {
        XCTAssertTrue(MemorySecretPIIGate.passesLuhn("4111111111111111"))
        XCTAssertTrue(MemorySecretPIIGate.passesLuhn("4111 1111 1111 1111"))
        XCTAssertFalse(MemorySecretPIIGate.passesLuhn("1234567890123456"))
        XCTAssertFalse(MemorySecretPIIGate.passesLuhn("4111"))            // too short
        XCTAssertFalse(MemorySecretPIIGate.passesLuhn(String(repeating: "1", count: 20))) // too long
    }

    func testTenDigitPhoneIsNotMisclassifiedAsCreditCard() {
        // A US phone has 10 digits; the card regex needs 13-19, and Luhn guards the
        // rest. The phone must surface as PII phone, never as a card.
        let findings = MemorySecretPIIGate.findings(in: "call 312-555-0199 after noon")
        XCTAssertTrue(findings.map(\.id).contains("us-phone-number"))
        XCTAssertFalse(findings.map(\.id).contains("credit-card-number"))
    }

    // MARK: - IPv4 octet bounds (must-fix #3)

    func testValidIPv4IsDetected() {
        XCTAssertEqual(MemorySecretPIIGate.findings(in: "host 192.168.20.15").map(\.id), ["ipv4-address"])
    }

    func testOutOfRangeIPv4IsPreservedNotDeleted() {
        // 999.888.777.666 is not a real address; deleting it as PII would corrupt a
        // legitimate version-like string.
        XCTAssertEqual(MemorySecretPIIGate.evaluate("ver 999.888.777.666 ok", policy: .reject), .allow)
    }

    func testIPv4OctetHelperDirectly() {
        XCTAssertTrue(MemorySecretPIIGate.hasBoundedIPv4Octets("0.0.0.0"))
        XCTAssertTrue(MemorySecretPIIGate.hasBoundedIPv4Octets("255.255.255.255"))
        XCTAssertFalse(MemorySecretPIIGate.hasBoundedIPv4Octets("256.1.1.1"))
        XCTAssertFalse(MemorySecretPIIGate.hasBoundedIPv4Octets("1.2.3"))      // too few octets
        XCTAssertFalse(MemorySecretPIIGate.hasBoundedIPv4Octets("1.2.3.4.5"))  // too many octets
    }

    // MARK: - Entropy + decode rescan

    func testHighEntropyTokenIsDetected() {
        let token = ["Az9qLm8Pr2Vx7", "Ns4Tu6Wy1Za3", "Qb5Cd7Ef9Gh2", "Jk4Mn6"].joined()
        let findings = MemorySecretPIIGate.findings(in: "blob \(token)")
        XCTAssertEqual(findings.map(\.id), [MemorySecretPIIGate.highEntropyFindingID])
        XCTAssertEqual(findings.first?.kind, .secret)
    }

    func testBase64EncodedSecretIsDetectedViaDecodeRescan() {
        let encoded = Data(("ghp_" + "A1B2C3D4E5F6G7H8I9J0K1L2M3N4O5P6Q7R8").utf8).base64EncodedString()
        let findings = MemorySecretPIIGate.findings(in: "payload \(encoded)")
        XCTAssertTrue(
            findings.map(\.id).contains("github-token"),
            "Expected the decoded github token to surface; got \(findings.map(\.id))"
        )
    }

    func testPrivateKeyMarkerFragmentIsRejected() {
        let marker = ["-----BEGIN", "OPENSSH", "PRIVATE", "KEY-----"].joined(separator: " ")
        let body = """
        setup note
        \(marker)
        """
        let verdict = MemorySecretPIIGate.evaluate(body, policy: .reject)
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected reject for private-key marker fragment, got \(verdict)")
            return
        }
        XCTAssertEqual(findings.map(\.id), ["private-key-marker"])
        XCTAssertEqual(findings.first?.kind, .secret)
    }

    func testDecodedOpenSSHPrivateKeyPayloadIsRejected() {
        let magic = "openssh-key-v1"
        let payload = Data(Array(magic.utf8) + [0x00, 0xFF, 0xFE, 0x80] + Array(repeating: 0x41, count: 32))
        let encoded = payload.base64EncodedString()
        let verdict = MemorySecretPIIGate.evaluate("payload \(encoded)", policy: .reject)
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected reject for decoded private-key payload marker, got \(verdict)")
            return
        }
        XCTAssertEqual(findings.map(\.id), ["openssh-private-key-payload"])
        XCTAssertEqual(findings.first?.kind, .secret)
    }

    func testOpenSSHFormatStringWithoutDecodedNulIsAllowed() {
        let formatMarker = ["openssh", "key", "v1"].joined(separator: "-")
        XCTAssertEqual(
            MemorySecretPIIGate.evaluate("parser handles \(formatMarker) payloads", policy: .reject),
            .allow
        )
    }

    func testPublicKeyMarkerIsAllowed() {
        let marker = ["-----BEGIN", "PUBLIC", "KEY-----"].joined(separator: " ")
        XCTAssertEqual(
            MemorySecretPIIGate.evaluate(marker, policy: .reject),
            .allow
        )
    }

    // MARK: - Redaction (must-fix #4): bounded, fail-closed downgrade

    func testRedactRemovesLocatableSecretSpan() {
        let ssn = "123-45-6789"
        let verdict = MemorySecretPIIGate.evaluate("My ssn is \(ssn) keep the rest.", policy: .redact)
        guard case .redact(let text, let findings) = verdict else {
            XCTFail("Expected redact for a locatable PII span, got \(verdict)")
            return
        }
        XCTAssertFalse(text.contains(ssn), "Redacted text still contains the SSN.")
        XCTAssertTrue(text.contains("[REDACTED-PII]"))
        XCTAssertTrue(text.contains("keep the rest."))
        XCTAssertEqual(findings.map(\.id), ["us-ssn"])
    }

    func testRedactDowngradesToRejectForNonLocatableDecodedSecret() {
        // The secret only exists in a base64-decoded view — there is no span to
        // excise on the original body, so redact MUST fail closed to reject.
        let encoded = Data(("ghp_" + "A1B2C3D4E5F6G7H8I9J0K1L2M3N4O5P6Q7R8").utf8).base64EncodedString()
        let verdict = MemorySecretPIIGate.evaluate("payload \(encoded)", policy: .redact)
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected reject (downgrade) for non-locatable decoded secret, got \(verdict)")
            return
        }
        XCTAssertTrue(findings.map(\.id).contains("github-token"))
    }

    func testRedactRemovesMultipleLocatableSpans() {
        // Exercises the right-to-left, offset-recomputed multi-span path.
        let ssn = "123-45-6789"
        let ip = "192.168.20.15"
        let verdict = MemorySecretPIIGate.evaluate(
            "ssn \(ssn) and host \(ip) remain readable around them.",
            policy: .redact
        )
        guard case .redact(let text, let findings) = verdict else {
            XCTFail("Expected redact for two locatable PII spans, got \(verdict)")
            return
        }
        XCTAssertFalse(text.contains(ssn))
        XCTAssertFalse(text.contains(ip))
        XCTAssertTrue(text.contains("[REDACTED-PII]"))
        XCTAssertTrue(text.contains("remain readable around them."))
        XCTAssertEqual(Set(findings.map(\.id)), ["us-ssn", "ipv4-address"])
    }

    func testRedactCleanTextIsAllow() {
        XCTAssertEqual(MemorySecretPIIGate.evaluate("totally fine note", policy: .redact), .allow)
    }

    // MARK: - Finding-id / label contract carriage (must-fix #1)

    func testFindingCarriesBothIDAndLabel() {
        let body = "key " + ["sk", "ant", String(repeating: "z", count: 20)].joined(separator: "-")
        XCTAssertEqual(MemorySecretPIIGate.findingIDs(in: body), ["anthropic-api-key"])
        XCTAssertEqual(MemorySecretPIIGate.labels(in: body), ["Anthropic API key detected"])
    }

    func testFindingsAreDedupedByID() {
        let key = ["sk", "ant", String(repeating: "a", count: 20)].joined(separator: "-")
        // Two occurrences of the same secret => one finding.
        let ids = MemorySecretPIIGate.findingIDs(in: "\(key) and again \(key)")
        XCTAssertEqual(ids, ["anthropic-api-key"])
    }

    // MARK: - Corpus unavailable: fail-closed (G7 gate audit finding)
    //
    // The gate design contract states: when the secret-pattern corpus is missing
    // or fails to load, `evaluate` MUST return `.reject` with a synthetic
    // `corpus-unavailable` finding — it must NEVER fall through to `.allow`.
    // These tests exercise that branch via the `_evaluate(_:policy:overrideCorpus:)`
    // testing seam, which injects `LoadedCorpus.unavailable` so the test runs
    // without depending on filesystem state.

    func testCorpusUnavailableRejectsEverything_benignText() {
        // Even completely benign text must be rejected when the corpus is missing.
        let verdict = MemorySecretPIIGate._evaluate(
            "I prefer dark mode and tabs over spaces.",
            policy: .reject,
            overrideCorpus: .unavailable
        )
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected .reject when corpus is unavailable, got \(verdict)")
            return
        }
        XCTAssertEqual(
            findings.map(\.id),
            [MemorySecretPIIGate.corpusUnavailableFindingID],
            "Fail-closed finding must carry the stable corpus-unavailable id."
        )
        XCTAssertEqual(
            findings.first?.label,
            MemorySecretPIIGate.corpusUnavailableLabel,
            "Fail-closed finding must carry the daemon-contract label verbatim."
        )
        XCTAssertEqual(
            findings.first?.kind,
            .secret,
            "Corpus-unavailable finding kind must be .secret (existing audit contract)."
        )
    }

    func testCorpusUnavailableRejectsEverything_emptyText() {
        // Even the empty string must be rejected — the gate cannot verify
        // anything without a corpus.
        let verdict = MemorySecretPIIGate._evaluate(
            "",
            policy: .reject,
            overrideCorpus: .unavailable
        )
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected .reject for empty text when corpus is unavailable, got \(verdict)")
            return
        }
        XCTAssertEqual(findings.map(\.id), [MemorySecretPIIGate.corpusUnavailableFindingID])
    }

    func testCorpusUnavailableRejectsUnderRedactPolicy() {
        // The `.redact` policy must ALSO fail closed when the corpus is unavailable;
        // the policy default of `.reject` must not be the only guarded branch.
        let verdict = MemorySecretPIIGate._evaluate(
            "totally fine note",
            policy: .redact,
            overrideCorpus: .unavailable
        )
        guard case .reject(let findings) = verdict else {
            XCTFail("Expected .reject under .redact policy when corpus is unavailable, got \(verdict)")
            return
        }
        XCTAssertEqual(findings.map(\.id), [MemorySecretPIIGate.corpusUnavailableFindingID])
    }

    func testAvailableCorpusStillAllowsBenignText() {
        // Sanity-check: when we do NOT override the corpus (real production path),
        // benign text is still allowed. This pair-test distinguishes "test
        // exercises the unavailable path" from "the gate is just broken globally".
        XCTAssertTrue(
            MemorySecretPIIGate.isAvailable,
            "Production corpus must be available; if not, ALL tests in this class are vacuous."
        )
        let verdict = MemorySecretPIIGate._evaluate(
            "I prefer dark mode and tabs over spaces.",
            policy: .reject,
            overrideCorpus: corpus  // real loaded corpus
        )
        XCTAssertEqual(verdict, .allow, "Benign text must be .allow when corpus is available.")
    }
}

// MARK: - Test-only corpus accessor

// Constructs a minimal available-sentinel `LoadedCorpus` for use in
// `testAvailableCorpusStillAllowsBenignText`. The corpus payload is empty
// (zero patterns) because `_evaluate` delegates to `MemorySecretPIIGate.evaluate`
// (the real production scanner) whenever `overrideCorpus.available == true`, so
// only the `available` flag matters here.
private var corpus: MemorySecretPIIGate.LoadedCorpus {
    MemorySecretPIIGate.LoadedCorpus(
        version: "test-sentinel",
        patterns: [],
        entropy: nil,
        hexEntropy: nil,
        decoding: nil,
        available: true
    )
}

File: windows/tests/memory-search/OpenBurnBar.App.MemorySearch.Tests/MemorySecretPIIGateTests.cs
using System;
using System.Linq;
using OpenBurnBar.App.MemorySearch.Memory;
using Xunit;

namespace OpenBurnBar.App.MemorySearch.Tests;

/// <summary>
/// The corpus-driven secret/PII gate. Swift: <c>MemorySecretPIIGate</c>. Covers the committed-corpus
/// load, explicit secret/PII patterns, the Luhn + IPv4 validators, entropy heuristics, redaction,
/// derived-view decoding, and the fail-closed invariant (injected unavailable corpus).
/// </summary>
public sealed class MemorySecretPIIGateTests
{
    private static readonly MemorySecretPIIGate Gate = MemorySecretPIIGate.Shared;

    [Fact]
    public void EmbeddedCorpus_LoadsAndReportsVersion()
    {
        Assert.True(Gate.IsAvailable);
        Assert.Equal("openburnbar-project-code-secret-corpus-v5", Gate.CorpusVersion);
    }

    [Fact]
    public void BenignText_IsAllowed()
    {
        Assert.Equal(MemoryGateVerdict.Kind.Allow, Gate.Evaluate("The user prefers dark mode and two-space indents.").Decision);
    }

    [Theory]
    // Synthetic secret VECTORS that exercise the PII gate's detectors — not real
    // credentials (openai/anthropic keys are fake sequences, AKIA…EXAMPLE is the
    // AWS docs sample, the JWT is a hand-forged token). gitleaks:allow on each.
    [InlineData("my key is sk-abcdefghijklmnopqrstuvwxyz012345", "openai-api-key")] // gitleaks:allow
    [InlineData("token sk-ant-abcdefghijklmnop0123", "anthropic-api-key")] // gitleaks:allow
    [InlineData("ghp_abcdefghijklmnopqrstuvwxyz0123456789", "github-token")] // gitleaks:allow
    [InlineData("AKIAIOSFODNN7EXAMPLE here", "aws-access-key")] // gitleaks:allow
    [InlineData("jwt eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0.dozjgNryP4J3jVmNHl0w5N", "jwt")] // gitleaks:allow
    public void KnownSecrets_AreRejectedWithExpectedFindingId(string text, string expectedId)
    {
        var verdict = Gate.Evaluate(text);
        Assert.Equal(MemoryGateVerdict.Kind.Reject, verdict.Decision);
        Assert.Contains(expectedId, verdict.Findings.Select(f => f.Id));
    }

    [Fact]
    public void PrivateKeyBlock_IsRejected()
    {
        const string text = "-----BEGIN RSA PRIVATE KEY-----\nMIIB...\n-----END RSA PRIVATE KEY-----";
        Assert.Equal(MemoryGateVerdict.Kind.Reject, Gate.Evaluate(text).Decision);
    }

    [Fact]
    public void Email_And_Ssn_AreRejectedAsPii()
    {
        Assert.Contains("email-address", Gate.FindingIds("reach me at alice@example.com"));
        Assert.Contains("us-ssn", Gate.FindingIds("ssn 123-45-6789"));
    }

    [Fact]
    public void CreditCard_LuhnValidatorGatesTheMatch()
    {
        // 4111 1111 1111 1111 is a valid Luhn test card → rejected.
        Assert.Contains("credit-card-number", Gate.FindingIds("card 4111111111111111"));
        // A 16-digit run that fails Luhn is NOT a finding (data-integrity guard).
        Assert.DoesNotContain("credit-card-number", Gate.FindingIds("order 1234567812345678"));
    }

    [Fact]
    public void Ipv4_OctetValidatorGatesTheMatch()
    {
        Assert.Contains("ipv4-address", Gate.FindingIds("server at 192.168.1.100"));
        // Out-of-range octets (a version string) must NOT be flagged.
        Assert.DoesNotContain("ipv4-address", Gate.FindingIds("version 999.888.777.666"));
    }

    [Fact]
    public void HighEntropyToken_IsRejected()
    {
        // A mixed-charset 40-char token that matches no explicit pattern → general entropy path.
        var ids = Gate.FindingIds("secret value aB3xK9mZ2qW7rT1yU5pL8nD4vC6bH0jF2gS3dA1e");
        Assert.Contains("high-entropy-token", ids);
    }

    [Fact]
    public void HighEntropyHexToken_IsRejected()
    {
        // A 40-char hex string caps below the general 4.2 threshold but trips the hex path (>= 3.0).
        var ids = Gate.FindingIds("digest a94a8fe5ccb19ba61c4c0873d391e987982fbbd3");
        Assert.Contains("high-entropy-hex-token", ids);
    }

    [Fact]
    public void Redact_RemovesLocatablePii_AndReGatesClean()
    {
        var verdict = Gate.Evaluate("email me at alice@example.com please", MemoryGatePolicy.Redact);
        Assert.Equal(MemoryGateVerdict.Kind.Redact, verdict.Decision);
        Assert.NotNull(verdict.RedactedText);
        Assert.DoesNotContain("alice@example.com", verdict.RedactedText!);
        Assert.Contains("[REDACTED-PII]", verdict.RedactedText!);
    }

    [Fact]
    public void Base64EncodedSecret_IsCaughtViaDerivedView()
    {
        // base64("sk-abcdefghijklmnopqrstuvwxyz012345") — the raw string hides the key; the decoded
        // view surfaces it (non-locatable), so the gate still rejects.
        string encoded = Convert.ToBase64String(
            System.Text.Encoding.ASCII.GetBytes("sk-abcdefghijklmnopqrstuvwxyz012345"));
        var verdict = Gate.Evaluate("blob " + encoded);
        Assert.Equal(MemoryGateVerdict.Kind.Reject, verdict.Decision);
        Assert.Contains("openai-api-key", verdict.Findings.Select(f => f.Id));
    }

    [Fact]
    public void Labels_And_FindingIds_AreDedupedAndOrderPreserving()
    {
        const string text = "keys sk-abcdefghijklmnopqrstuvwxyz012345 and sk-zzzzzzzzzzzzzzzzzzzzzzzz9999";
        Assert.Equal(new[] { "openai-api-key" }, Gate.FindingIds(text));
        Assert.Equal(new[] { "OpenAI API key detected" }, Gate.Labels(text));
    }

    [Fact]
    public void UnavailableCorpus_FailsClosed_RejectingEverything()
    {
        var gate = new MemorySecretPIIGate(MemoryGateCorpus.Unavailable);
        Assert.False(gate.IsAvailable);

        var verdict = gate.Evaluate("totally benign text");
        Assert.Equal(MemoryGateVerdict.Kind.Reject, verdict.Decision);
        Assert.Equal(MemorySecretPIIGate.CorpusUnavailableFindingId, Assert.Single(verdict.Findings).Id);

        Assert.Equal(new[] { MemorySecretPIIGate.CorpusUnavailableFindingId }, gate.FindingIds("x"));
        Assert.Equal(new[] { MemorySecretPIIGate.CorpusUnavailableLabel }, gate.Labels("x"));
    }

    [Fact]
    public void CorruptCorpusJson_LoadsAsUnavailable()
    {
        Assert.False(MemoryGateCorpus.FromJson("{ not valid").Available);
        // Missing required pattern field → whole corpus unavailable (fail-closed).
        Assert.False(MemoryGateCorpus.FromJson("{\"version\":\"v\",\"patterns\":[{\"id\":\"x\"}]}").Available);
    }

    [Fact]
    public void CustomCorpus_IsInjectableAndDrivesTheGate()
    {
        const string json = "{\"version\":\"custom-v1\",\"patterns\":[" +
            "{\"id\":\"forbidden-word\",\"label\":\"Forbidden word\",\"kind\":\"secret\",\"regex\":\"FORBIDDEN\"}]}";
        var gate = new MemorySecretPIIGate(MemoryGateCorpus.FromJson(json));

        Assert.True(gate.IsAvailable);
        Assert.Equal("custom-v1", gate.CorpusVersion);
        Assert.Equal(MemoryGateVerdict.Kind.Reject, gate.Evaluate("this is FORBIDDEN").Decision);
        Assert.Equal(MemoryGateVerdict.Kind.Allow, gate.Evaluate("this is fine").Decision);
    }

    [Fact]
    public void Validators_AreCorrect()
    {
        Assert.True(MemoryGateValidators.PassesLuhn("4111111111111111"));
        Assert.False(MemoryGateValidators.PassesLuhn("1234567812345678"));
        Assert.False(MemoryGateValidators.PassesLuhn("411111")); // too short
        Assert.True(MemoryGateValidators.HasBoundedIpv4Octets("0.0.0.0"));
        Assert.True(MemoryGateValidators.HasBoundedIpv4Octets("255.255.255.255"));
        Assert.False(MemoryGateValidators.HasBoundedIpv4Octets("256.1.1.1"));
        Assert.False(MemoryGateValidators.HasBoundedIpv4Octets("1.2.3"));
    }

    [Fact]
    public void ShannonEntropy_IsZeroForEmpty_AndPositiveForMixed()
    {
        Assert.Equal(0, MemorySecretPIIGate.ShannonEntropy(""));
        Assert.True(MemorySecretPIIGate.ShannonEntropy("abcdefghij") > 3.0);
    }
}
